Add Dynamic Mock support

haikusw opened the pull request:
Sometimes you want the same URL to return different results at different times during a test - for example, an error payload first and then a success payload.

Dynamic Mock support enables this by allowing one to setup a closure (`onMockFor`) that gets called as a first step in mock matching (in Mocker's `mock(for request: URLRequest) -> Mock?`). If this closure returns a Mock, that is the mock returned from `mock(for …`) and is used to fulfill the mocked network request.

----

The particular use case I had was for a service call made with an expired token. The networking layer in this app will catch the expired token error, automatically make a refresh token service call, and then auto-retry the original service call.

In this situation I needed to have the mock for the original service call return a 401 for the expired token as long as the currentToken doesn't match the newGoodToken returned by the mocked refresh token call (which is auto-saved into currentToken by the networking layer).

So my (simplified) `onMockFor` will look something like this:


```
let refreshMock = Mock(url: refreshURL, dataType: .json, statusCode: 200, data: [.get: goodRefreshTokenData])
refreshMock.register()
let failureMock = Mock(url: serviceURL, dataType: .json, statusCode: 401, data: [.get, Data()])
failureMock.register()

let successMock = Mock(url: serviceURL, dataType: .json, statusCode: 200, data: [.get: goodMockData])
// Note: do not register! Will overwrite failureMock (same URL).

// Then add the `onMockFor` closure to check if we have the new token
Mocker.onMockFor = { (request: URLRequest) -> Mock? in
  guard request.url == serviceURL else { return nil }
  if currentToken == newGoodToken {
      return successMock
  }
  return nil
}
```
----

Seemed like it might be useful to others, and I'd rather not maintain our own custom fork of Mocker separate from yours, so I thought I'd open a PR for this feature in case you'd like to incorporate it in Mocker proper.  Tried to stick to your format and style and the majority of the code added is tests.  Happy to change things to make it work better for you.

haikusw committed:
DynamicMocks: add onMockFor to provide for dynamic mock injection

Sometimes you want a url to return different things for the same URL during a test (error and then success, for example).  Dynamic Mock support enables that.

haikusw committed:
DynamicMocks - make onMockFor accessible without @testable import

haikusw committed:
Dynamic Mocks: fix indentation to match project settings.

(tabs to spaces)

haikusw committed:
Dynamic Mocks: improve comment

/// comments internal to a function do not show up in Quick Help so move that information to the function comment and demote internal comments to //

haikusw commented:
I'm not sure what the Bitrise failure is about.  I don't see that it even tried to do anything?

Boris-Em commented on MockerTests/MockerTests.swift:
```diff
@@ -401,4 +408,126 @@ final class MockerTests: XCTestCase {
+        wait(for: [expectation], timeout: 2.0)
+    }
+    
+    /// `onMockFor` returning nil for non-matching replacement mock shouldn't change previous behavior
```
We try to write all of our tests following the "It should..." format.
If this suggestion makes sense feels free to commit it directly.
```suggestion
    /// It should not change previous behavior when `onMockFor` returns `nil`.
```

Boris-Em commented on MockerTests/MockerTests.swift:
```diff
@@ -401,4 +408,126 @@ final class MockerTests: XCTestCase {
+        
+        let expectSuccess = expectation(description: "call url again - should succeed")
+        
+        // This retry should succeed becuase problemFixed is now true
```
```suggestion
        // This retry should succeed because `problemFixed` is now `true`
```

haikusw replied:
I'm happy to make this change but thought I'd mention that the reason I don't generally use backticks in internal quotes is because they are not displayed in Quick Help and so are only // style comments where backticks don't create a style change anywhere that I'm aware of.

I'm curious if there's a place internal quotes show up for which the styling elements are thus useful because if so I'll need to change my conventions.

haikusw replied:
I guess it's useful to clarify when the comment is referring to symbols or not even if it's not displayed differently.

AvdLee replied:
I guess if you make use of three slashes it should work `///`

haikusw replied:
I do not see that comments within the body of a function are pulled into the Quick Help for the function (Xcode 11.5).  If they appear somewhere else then I'd like to know about that.

In discussion with my colleagues we decided that even if it doesn't show up it helps the reader know you are referring to a symbol.

AvdLee replied:
Ah, I didn't see that this was an inline comment. Indeed, those are not showing in Quick Help but I agree that it still helps the reader!

Boris-Em commented on MockerTests/MockerTests.swift:
```diff
@@ -401,4 +408,126 @@ final class MockerTests: XCTestCase {
+        wait(for: [expectation], timeout: 2.0)
+    }
+    
+    /// `onMockFor` can change the Mock used for a url for subsequent fetches during same test
```
```suggestion
    /// It should call `onMockFor` to change the mock used for a url for subsequent fetches during the same test
    /// `onMockFor` can change the Mock used for a url for subsequent fetches during same test
```

Boris-Em commented on Sources/Mocker.swift:
```diff
@@ -23,6 +23,17 @@ public struct Mocker {
     public static var httpVersion: HTTPVersion = HTTPVersion.http1_1
     
+    /// Allow for dynamic Mocks.
+    /// If closure is set and a Mock is returned, that mock will override
```
```suggestion
    /// If the closure is set and a Mock is returned, that mock will override
```

Boris-Em reviewed (changes requested):
Great job @haikusw! 
This is a really great addition, thanks for contributing.

I have a few nits mostly related to documentation.
Maybe it would also be a good idea to udpate the readme to have a short section explaining the usage of this feature?

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw committed:
Update Sources/Mocker.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw commented:
@Boris-Em Thank you for the review. I'm glad you also see this as a useful addition.

I appreciate the suggestions for refinements to the comments, and I committed all your suggestions. Thanks.

haikusw committed:
Update MockerTests/MockerTests.swift

Co-authored-by: Boris Emorine <boris.emorine@gmail.com>

haikusw commented:
Ah, forgot the readme!  I'll have a go at that tomorrow.

haikusw committed:
Dynamic Mock: Update Readme as suggested

haikusw committed:
Merge branch 'dynamic-mock' of https://github.com/haikusw/Mocker into dynamic-mock

AvdLee commented on MockerTests/MockerTests.swift:
```diff
@@ -20,6 +20,13 @@ final class MockerTests: XCTestCase {
+    override func setUp() {
+        super.setUp()
+        
+        // make sure this is cleared before each test to avoid bleed-over between tests.
```
Nit: 
```suggestion
        // Make sure this is cleared before each test to avoid bleed-over between tests.
```

ghost commented:
<!--
  1 failure:  Failed to get the...
  2 warning:  Consider using fi..., Consider using fi...
  1 messages
  
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Failed to get the coverage - Error: The data couldn’t be read because it isn’t in the correct format.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  **MockerTests/DynamicMocksTests.swift#L11** - Consider using final for this class or use a struct (final_class)
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  **MockerTests/DynamicMocksTests.swift#L216** - Consider using final for this class or use a struct (final_class)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  View more details on <a href="https://app.bitrise.io/build/23201f39685e0c2f" target="_blank">Bitrise</a>
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against 5352902070d41bc999f437e06c23898af0f5992e
</p>

AvdLee reviewed (commented):
Hey Tyler,

Thanks a lot for this contribution. The problem it's solving is something common and definitely worth being able to Mock!

However, it's already possible without your implementation to test this. Let me explain this!

In your example, you'd like to make sure that the failure `Mock` is called before the `Success` Mock. You also want to make sure that the `Success` Mock is only called after the refresh `Mock`. 

This is something that is already possible with the framework by making use of the `Mock.onRequestExpectation` property and using the `enforceOrder` on `waitForExpectations`. 

```swift
wait(for: [failureMockExpectation, refreshMockExpectation, successMockExpectation], timeout: 2.0, enforceOrder: true)
```

You can combine this with the `onRequest` property.

```swift
refreshMock.onRequest { _, _ in
    successMock.register() // Overrides the previous failure Mock which is already called
}
```

Therefore, despite the hard work you've put into this, I think we can close this PR as the problem you're trying to solve can already be tested with existing logic.

Let me know your thoughts, I'm happy to discuss further!

haikusw commented:
@AvdLee Interesting. I tried using `onRequest` but it is called after the `Mocker.mock(for: request)` in `MockingURLProtocol startLoading()` - which makes sense since it is associated with the Mock itself, but that means it's too late to change the current mock.

Combining `onRequest` with `onRequestExpectation` didn't occur to me, so I'll have to try that. 

Actually, reading the code in `finishRequest`, I now think that using the `completion` handler for Mock for the tokenRefresh Mock will allow me to do exactly what I need much more simply. (How did I miss that? LOL).

Thanks for the pushback/guidance!

AvdLee commented:
@haikusw you're welcome! The best way to learn is by trying 💪🏻

Let me know if you have any other questions!

haikusw committed:
Make an example test and service to test to show the reason for onMockFor

This commit contains a test file with a simplified example of a service one wants to test and shows how onMockFor is used to test it.

I was not able to use existing Mocker functionality to create the test to test this functionality.  Might be a way, but I’m not seeing it. Hence onMockFor which, as you can see in this test, works perfectly.

haikusw committed:
convert tabs to spaces

Sorry - current client project uses tabs so that’s Xcode’s default on my machine currently.

haikusw commented:
Hi @AvdLee. I was not able to figure out a way to use existing Mocker functionality to write a test that would adequately test the service code in question.  

I have added a simplified version of some service code, and a test to test it, so that you can see something closer to what I need to write tests for.  I'm hoping that either you can let me know how I can write the test with existing Mocker functionality, or see why `onMockFor` is a needed addition.

The main issue with existing Mocker functionality in this situation seems to be that the existing hooks are not called at the right time to check that the processing done by the refresh token service call implementation on the mocked data it receives has actually worked before the mock for the second profile call (the retry) has been selected by MockingURLProtocol.

Thank you.

haikusw commented:
Hmm. My commits didn't get added to this PR though they are on the branch. Perhaps because this PR is closed.  Not sure if I should open another or if you re-opening this one will update it to the current code on the [branch](https://github.com/haikusw/Mocker/tree/dynamic-mock)?  Happy to do whatever is best.

haikusw commented:
Thought I'd note that I thought of at least two other approaches to adding the needed functionality but the method in this branch was the smallest change I could make to the Mocker source that would satisfy the requirements (~8 lines of actual code, besides that in tests).

The example in the ReadMe feels heavier than I'd like though. I found creating a small example showing both how to use this advanced functionality and motivating when it's needed to be challenging.

haikusw commented:
Tried implementing the simpler of the other approaches (making the dictionary of data be a dictionary of { () -> Data } closures instead, but that doesn't let you dynamically change the httpStatus (from error to success for example) or the headers, so doesn't really work.  Was also source-breaking which seemed like a non-starter and why I didn't try it previously. So this PR still seems like the cleanest way to add the necessary functionality.

AvdLee commented:
@haikusw can you change your unit test in the branch you linked to the code I suggested? I'll review that and let you know what I think. I'm pretty sure my suggested approach should work.

- By the time the `onRequest` is called, you still have time to register and overwrite a new mock. 
- The token is refreshed
- The new mock is called
- Second mock expectation succeeds

haikusw committed:
Add my understanding of the two Tests AvdLee suggested

Might be mis-understanding so it seemed best to write my understanding in code to aid communication.

haikusw committed:
Set project settings to spaces to match project convention

Changed the project setting to spaces to match the convention for the project so that I don’t keep accidentally submitting tabs (my Xcode default seting because of current client requirements).  This will help other contributors avoid the same issue.

haikusw commented:
Hi!  Thanks for the reply.  I'm happy to, but I think we may be misunderstanding one another:

Which Mock in the DynamicMocksTests.swift `testOnMockForConditionalReplacement` test do you want me to attach the `onRequest` handler to?

--

Looking at the DynamicMocksTests.swift file:

If I attach the `onRequest` handler to the `refreshTokenMock` then it is executed _before_ the code that receives the data from that mock runs - and that code is part of what this test is testing: 

- that it properly decodes the mocked refresh payload
- saves the token therein (to the fake keychain)
- returns success to the network layer code
- which then retries the original network call that failed due to `failingProfileMock`.

If I attach the `onRequest` handler to the `failingProfileMock` so that this refresh payload processing will already be done the second time the `failingProfileMock` is matched, then it is too late to swap in the `succeedingProfileMock` for the failing one.  This is because the `onRequest` handler is part of the `failingProfileMock` which has already been matched and is being executed at the time its `onRequest` handler is called (line 46 MockingURLProtocol.swift, _after_ the `Mocker.mock(for: request)` on line 37).

(To make sure I was reading the code correctly and to aid communication, I just coded up my understanding of your suggestion in both of the variations I thought through above, and both tests fail for the reasons I expected above.  I've committed those tests with a "FAIL_" prefix to the function names so that all enabled tests still pass CI builds. This way you can see the code representing my understanding of your suggestions, try it out if you wish, and/or clarify your suggestion).

Thank you.

haikusw commented:
(if you re-open this PR then all the new commits will show up here)

AvdLee commented on MockerTests/DynamicMocksTests.swift:
```diff
@@ -0,0 +1,294 @@
+
+        failingProfileMock.onRequest = { _, _ in
+            // this won't help because it's too late.
+            if FakeService.accessToken.token == "good token" {
```
I've checked out your branch and commented out this code. Then the test succeeded.

You have to register the succeeding mock as soon as it failed once.

haikusw replied:
Hi @AvdLee - thanks for taking a look at it.

Yes.  Unfortunately, commenting out that line removes testing steps 1 & 2 of the four aspects of the refresh portion of the code under test that I listed above.  That's a problem because if the dynamic token refresh handling code fails to process the mocked refresh token payload properly and deliver the results to the right place(s), then the test will still pass. 

It is not possible to test this in Mocker without some way to dynamically replace a mock when Mocker is looking up which mock to use because there's no place else to test that the chained network call did it's work properly (since it's chained control doesn't return to the test code itself during the chained processing).

This simple example code for this sample test doesn't make it look like there's much to do, but in the real world it's more complicated and something that really wants to be tested (refreshing the accessToken with a refreshToken can fail because the refreshToken has expired; then you have to get a new refreshToken, which then allows you to get a new accessToken, etc).

haikusw replied:
This is a specific instance of the general limitation which the `onMockFor` addition addresses: there is currently no way to dynamically set a mock after processing the payload of a mocked network call within a series of chained networking calls when two of the calls call the same url (well, possible if there are two or more network calls between the repeated call to the same URL, but not if is only 1).

Imagine testing a multi-step combine chain which makes repeated calls to a network endpoint and you want to change the mock for each time you call the endpoint but you want/need to test the results of the combine chain operators between each call to the endpoint to figure out which mock to activate - that's another example of the use case for `onMockFor`.

AvdLee replied:
Correct me if I'm wrong but don't you want to simply verify that the correct new token is sent with the second succeeding request?

If so, you can do this with the `onRequest` of the succeeding mock and assert in there

haikusw replied:
Interesting idea!  In this simplified example test situation that is indeed all that needs to be tested. I modified the test and it works as you suggest.  Nice!

I'll have to go dig into my real-world shipping-code use-case and see if something similar can work there.  It's a quite a bit more complicated; I have variants on the tests that test the error handling paths via mocks that simulate the intermediate service requests returning various errors and making sure they either bubble up through the combine pipeline to the original caller properly or trigger alternate auto-try-to-fix-and-retry handlers and then retry again.

Thanks for the feedback.

haikusw replied:
Tried it out in the real app tests and this approach does work for the sub-set of tests I tried it on there also (there are more, but some will take more work to rework).

While I do still think the `onMockfor` functionality will be useful in other situations, it's clearly not necessary for this one, and the no-changes-needed approach always wins :)

Thank you for your patience and your assistance helping me see an alternative no-changes-needed approach.

AvdLee replied:
All good, glad I could help! 💪 Thanks for staying motivated to integrate my feedback!

haikusw commented:
@AvdLee Hi.  Not sure if you get charged for WeTransferBot work, but either way it would probably be better to have it ignore closed PRs.  Cheers.

AvdLee commented:
@haikusw we're not, so no worries 🙂 

Are you continuing on this feature? If so, I'm happy to take another look and reopen.

haikusw commented:
cool.
No, thank you. You showed me a way to do what I needed to do without it and I don't have another use case at the moment so can't motivate adding it.  Thanks for asking though.
